cumulus/parachains/runtimes/constants/src/default_configs.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This builds on the DefaultConfigs for the various pallets to maximise the usefulness of
+// [`derive_impl`] for the system parachains runtimes for testnets.
+
+use super::*;
+use frame_support::pallet_prelude::*;
+use parachains_common::{AccountId, Balance, Hash, Nonce};
+use polkadot_runtime_common::BlockHashCount;
+
+/// A struct for which DefaultConfigs can be defined for pallets common to system parachains.
+pub struct SystemParachainDefaultConfig;
+
+/// [`frame_system::DefaultConfig`] for system parachains.
+#[frame_support::register_default_impl(SystemParachainDefaultConfig)]
+impl frame_system::DefaultConfig for SystemParachainDefaultConfig {
+	/// The default type for storing how many extrinsics an account has signed.
+	type Nonce = Nonce;
+	/// The default type for hashing blocks and tries.
+	type Hash = Hash;
+	/// The default hashing algorithm used.
+	type Hashing = sp_runtime::traits::BlakeTwo256;
+	/// The default identifier used to distinguish between accounts.
+	type AccountId = AccountId;
+	/// The lookup mechanism to get account ID from whatever is passed in dispatchers.
+	type Lookup = sp_runtime::traits::AccountIdLookup<Self::AccountId, ()>;
+	/// The maximum number of consumers allowed on a single account.
+	type MaxConsumers = frame_support::traits::ConstU32<16>;
+	/// The default data to be stored in an account.
+	type AccountData = pallet_balances::AccountData<Balance>;
+	/// What to do if a new account is created.
+	type OnNewAccount = ();
+	/// What to do if an account is fully reaped from the system.
+	type OnKilledAccount = ();
+	/// Weight information for the extrinsics of this pallet.
+	type SystemWeightInfo = ();
+	/// This is used as an identifier of the chain.
+	type SS58Prefix = ();
```

Comment:
Do we really need the parameters that are going to be overloaded in the impl for `Runtime`?
Whenever I need to change some config parameter, if I find it in default config, after I change it, it's good to have a confidence that it wont be overloaded, and I do not need to check every runtime config.

Reply:
Yes I initially wanted to do this, but we need every trait member that hasn't been declared as `no_default` in `pallet_system::Config`, but then if we do that they will have to be specified manually in testing and relays etc. We'd need the following types to be declared `no_default`
```
type SystemWeightInfo = ();
type SS58Prefix = ();
type Version = ();
type BlockWeights = ();
type DbWeight = ();
type OnSetCode = ();
```

I agree that it would be a better misconfiguration check if the compiler told us when something was left out unintentionally, even if the relays and tests had a few more unit types to fill in when they are configuring the pallet. I would support tagging these as `no_default`. WDYT @gupnik @kianenigma? Or is there a more elegant solution? For example the introduction of `no_default_except_tests` or something similar.

Reply:
My best advice is https://github.com/paritytech/polkadot-sdk/pull/4689#discussion_r1625651779. TLDR; Provide a default that will make the compiler happy, but will force downstream users to override it. I find what is being done here to be too risky. 

Option 2: remember that `#[pallet::config(with_default)]` is merely one way to generate `trait DefaultConfig`. We can use the underlying machinery and generate our own custom `trait DefaultConfig` and still be able to use `derive_impl` with it. 


@gupnik how far are we from being able to do this? If it is possible, where is it documented? If not, it is worth being noted in an issue, with an evaluation of how much work it is. 

I have other use cases in mind that I specualte will be useful, if we allow anyone to provide `trait DefaultConfig` for any pallet, and the `derive_impl` syntax to be able to pick any of them up.

Reply:
> We can use the underlying machinery and generate our own custom `trait DefaultConfig` and still be able to use `derive_impl` with it

Indeed! This is the way I imagine it would look like:

```rust
pub trait CustomDefaultConfigProvider {
	type RuntimeOrigin;
	type RuntimeCall;
	type RuntimeTask;
}

pub struct CustomDefaultConfig;

#[crate::register_default_impl(CustomDefaultConfig)]
impl CustomDefaultConfigProvider for CustomDefaultConfig {
	#[inject_runtime_type]
	type RuntimeOrigin = ();
	#[inject_runtime_type]
	type RuntimeCall = ();
	#[inject_runtime_type]
	type RuntimeTask = ();
}

struct Runtime;

#[derive_impl(CustomDefaultConfig, no_aggregated_types)
impl Config for Runtime {}
```

Will try this out and add as an example via docify if works.

Reply:
Added an issue for this: https://github.com/paritytech/polkadot-sdk/issues/4827

Reply:
> Will try this out and add as an example via docify if works

Did it work?

Reply:
> > Will try this out and add as an example via docify if works
> 
> Did it work?

It did! Just need to find some time to put it into a PR :)